tests/build/run_cython_not_in_cwd.srctree

```diff
@@ -0,0 +1,56 @@
+
+PYTHON check_paths_from_cython.py
+
+######## src/foo/__init__.py ########
+
+# foo is a package and foo.bar is the module to build
+
+######## src/foo/bar.pyx ########
+
+def func(x):
+    return 2*x
+
+######## build/src/foo/placeholder.txt ########
+
+# Build outputs go in the build directory and cython is run from here.
+
+######## check_paths_from_cython.py ########
+
+import sys
+import subprocess
+
+python = sys.executable
+
+def cython(cmdstr, cwd=None):
+    argv = [sys.executable, '-m', 'cython'] + cmdstr.split()
+    subprocess.run(argv, check=True, cwd=cwd)
```

Comment:
You can also use something like `CYTHON src/pkg/mod.pyx` (or the same with `CYTHONIZE`) as command at the top to do the build for you before you run the (then simpler) script.

Reply:
How would I set the working directory in that case?

The purpose of the test here is to run the same command from two different working directories.

I realise that cython has a `--working` CLI option but I didn't want to write the test with the assumption that setting that option would be equivalent to actually running cython from a different working directory.

Reply:
Ah, right. It seems something we might want to do in more than one test, so we could allow `cd` as a command (or maybe `CD`) and set the argument as the working directory for the following commands.

Could be a later cleanup PR, though.

Reply:
Ideally if the top part was just a shell script I would have written it like this:
```
cython src/pkg/mod.pyx -o build/src/pkg/mod1.pyx.c
cd build && cython ../src/pkg/mod.pyx -o src/pkg/mod1.pyx.c
python check.py
```
I suppose a simple version could be like:
```
CYTHON src/pkg/mod.pyx -o build/src/pkg/mod1.pyx.c
CD build
CYTHON ../src/pkg/mod.pyx -o src/pkg/mod1.pyx.c
CD ..
PYTHON check.py
```
Managing the changed directory in the test runner could be fiddly though. I would say it is not worth it unless there are many tests that need this.

Reply:
Here's an untested proposal:
```patch
diff --git a/runtests.py b/runtests.py
index 3b8abe9aa..5473e4f1a 100755
--- a/runtests.py
+++ b/runtests.py
@@ -2028,6 +2028,9 @@ class EndToEndTest(unittest.TestCase):
                     envvar = None
                 env.pop(envvar, None)
                 continue
+            elif command[0] == 'CD':
+                os.chdir(command[1] if len(command) > 1 else self.workdir)
+                continue
             time_category = 'etoe-build' if (
                 'setup.py' in command or 'cythonize.py' in command or 'cython.py' in command) else 'etoe-run'
             with self.stats.time('%s(%d)' % (self.name, command_no), 'c', time_category):
```

Reply:
Using chdir is almost always a bad idea. Instead the runner should maintain some notion of a working directory and then pass that as the `cwd` parameter to every single subprocess invocation. That is why I said that it could be fiddly.

Reply:
It's how things currently work, though. The EndToEnd tests have `chdir(workdir)` in their `setUp()` method and reset it in `tearDown()`.

Longer patch, also untested:
```patch
diff --git a/runtests.py b/runtests.py
index 3b8abe9aa..22aef7447 100755
--- a/runtests.py
+++ b/runtests.py
@@ -1996,8 +1996,6 @@ class EndToEndTest(unittest.TestCase):
     def setUp(self):
         from Cython.TestUtils import unpack_source_tree
         _, self.commands = unpack_source_tree(self.treefile, self.workdir, self.cython_root)
-        self.old_dir = os.getcwd()
-        os.chdir(self.workdir)
 
     def tearDown(self):
         if self.cleanup_workdir:
@@ -2008,7 +2006,6 @@ class EndToEndTest(unittest.TestCase):
                     time.sleep(0.1)
                 else:
                     break
-        os.chdir(self.old_dir)
 
     def runTest(self):
         self.success = False
@@ -2020,6 +2017,7 @@ class EndToEndTest(unittest.TestCase):
         cmd = []
         out = []
         err = []
+        workdir = self.workdir
         for command_no, command in enumerate(self.commands, 1):
             if command[0] == "UNSET":
                 try:
@@ -2028,15 +2026,21 @@ class EndToEndTest(unittest.TestCase):
                     envvar = None
                 env.pop(envvar, None)
                 continue
+            elif command[0] == "CD":
+                if len(command) == 1:
+                    workdir = self.workdir
+                else:
+                    workdir = os.path.join(workdir, command[1])
+                continue
             time_category = 'etoe-build' if (
                 'setup.py' in command or 'cythonize.py' in command or 'cython.py' in command) else 'etoe-run'
             with self.stats.time('%s(%d)' % (self.name, command_no), 'c', time_category):
                 if self.capture:
-                    p = subprocess.Popen(command, stderr=subprocess.PIPE, stdout=subprocess.PIPE, env=env)
+                    p = subprocess.Popen(command, stderr=subprocess.PIPE, stdout=subprocess.PIPE, env=env, cwd=workdir)
                     _out, _err = p.communicate()
                     res = p.returncode
                 else:
-                    p = subprocess.call(command, env=env)
+                    p = subprocess.call(command, env=env, cwd=workdir)
                     _out, _err = b'', b''
                     res = p
             cmd.append(command)
```

Reply:
That being said I don't want to mess around with the runner too much in this PR. It already uses chdir so why not use it more I guess.

I have just pushed this implementation of `CD`.

Reply:
Our messages crossed paths. I've pushed that longer version here now.

The CI won't run until someone approves it though.